bson/src/main/org/bson/codecs/pojo/LazyPropertyModelCodec.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.bson.codecs.pojo;
+
+import org.bson.BsonReader;
+import org.bson.BsonType;
+import org.bson.BsonWriter;
+import org.bson.codecs.Codec;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.RepresentationConfigurable;
+import org.bson.codecs.configuration.CodecConfigurationException;
+import org.bson.codecs.configuration.CodecRegistry;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.bson.codecs.pojo.PojoSpecializationHelper.specializeTypeData;
+
+class LazyPropertyModelCodec<T> implements Codec<T> {
```

Comment:
This is the main change - instead of the LazyPojoCodec - this makes the PropertyModel cached codec itself lazy and reduces the calls back to the registry that previously existed thanks to the PropertyCodecRegistry containing using the FallbackPropertyCodecProvider